src/framework/annotator_mgr.cpp

```diff
@@ -583,7 +576,185 @@ namespace uima {
       }
 
       UIMA_ANNOTATOR_TIMING(iv_clTimerLaunchProcess.stop());
-      return(utRetVal);
+      return utRetVal;
+    }
+
+    CAS *AnnotatorManager::processUntilNextOutputCas() {
+      unique_ptr<Flow> flow{};
+      while (true) {
+        CAS *currentCas = nullptr;
+        Step nextStep;
+        flow = nullptr;
+
+        // get a cas from the stack
+        if (casIterStack.empty())
+          return nullptr;
+
+        StackFrame &frame = casIterStack.top();
+        try {
+          if (frame.casMultiplier && frame.casMultiplier->hasNext()) {
+            currentCas = &frame.casMultiplier->next();
+            // compute flow for newly produced CAS
+            flow = frame.originalFlow->newCasProduced(*currentCas, frame.lastEngineKey);
+          }
+        } catch (Exception &exception) {
+          if (!frame.originalFlow->continueOnFailure(frame.lastEngineKey /* ,exception */))
+            throw;
+        }
+
+        if (!currentCas) {
+          // if there is no more output CASes from the stack, take the original CAS that was processed by
+          // the CAS Multiplier and continue with its flow
+          currentCas = frame.originalCas;
+          flow = std::move(frame.originalFlow);
+          currentCas->setCurrentComponentInfo(nullptr);
+          casIterStack.pop();
+        }
+
+        activeCASes.insert(currentCas);
```

Comment:
I don't see where you remove CASes from `activeCASes`. That will lead to dangling pointers. The `release()` method will segfault.

Reply:
I chose to release everything in the caller launchProcessDocument instead with a try catch block, because I thought this method is busy enough.

Reply:
Yes you're right, I will fix that.